File: strings/group-anagrams/group-anagrams.ts
function groupAnagrams(strs: string[]): string[][] {
  const n: number = strs.length;

  const map: Record<string, string[]> = {};
  const res: string[][] = [];

  for (let i = 0; i < n; i++) {
    const characters: string = strs[i];
    const key: string = characters.split('').sort().join('');

    if (!map[key]) {
      map[key] = [];
    }
    const list: string[] = map[key];
    list.push(strs[i]);
  }

  const values: string[][] = Object.values(map);

  for (let i = 0; i < values.length; i++) {
    res.push(values[i]);
  }

  return res;
}

function groupAnagrams2(strs: string[]): string[][] {
  const n: number = strs.length;
  const map: Record<string, string[]> = {};
  const res: string[][] = [];

  for (let i = 0; i < n; i++) {
    const freq: number[] = new Array(26).fill(0);
    const characters: string = strs[i];

    for (let i = 0; i < characters.length; i++) {
      freq[characters[i].charCodeAt(0) - 'a'.charCodeAt(0)]++;
    }

    const key: string = freq.join('#');

    if (!map[key]) {
      map[key] = [];
    }

    const list: string[] = map[key];
    list.push(strs[i]);
  }

  const values = Object.values(map);
  for (let i = 0; i < values.length; i++) {
    res.push(values[i]);
  }

  return res;
}

File: strings/group-anagrams/group-anagrams.go
package groupanagrams

import (
	"sort"
	"strings"
)

func groupAnagrams(strs []string) [][]string {
	mp := map[string][]string{}
	res := [][]string{}

	for _, str := range strs {
		characters := strings.Split(str, "")

		sort.Strings(characters)

		key := strings.Join(characters, "")

		if _, ok := mp[key]; !ok {
			mp[key] = []string{}
		}

		list := mp[key]
		mp[key] = append(list, str)
	}

	for _, list := range mp {
		res = append(res, list)
	}

	return res
}

func groupAnagrams2(strs []string) [][]string {
	mp := map[string][]string{}
	res := [][]string{}

	for _, str := range strs {
		charFreq := make([]int, 26)

		for _, c := range []byte(str) {
			charFreq[c-'a']++
		}

		keyBuilder := []byte{}

		for _, n := range charFreq {
			keyBuilder = append(keyBuilder, byte(n))
			keyBuilder = append(keyBuilder, '#')
		}

		key := string(keyBuilder)

		if _, ok := mp[key]; !ok {
			mp[key] = []string{}
		}

		list := mp[key]
		mp[key] = append(list, str)
	}

	for _, list := range mp {
		res = append(res, list)
	}

	return res
}
